keras/utils/data_utils.py

```diff
@@ -282,3 +295,295 @@ def validate_file(fpath, file_hash, algorithm='auto', chunk_size=65535):
+                self.queue.put(self.executor.apply_async(get_index, (self.dataset, i)), block=True)
+
+    def get(self):
+        """Create a generator to extract data from the queue.
+
+        #Returns
+            A generator
+        """
+        try:
+            while self.is_running():
+                yield self.queue.get(block=True).get()
+        except Exception as e:
+            self.stop()
+            raise StopIteration(e)
+
+    def stop(self, timeout=None):
+        """Stop running threads and wait for them to exit, if necessary.
+
+        Should be called by the same thread which called start().
+
+        # Arguments
+            timeout: maximum time to wait on thread.join()
+        """
+        self.stop_signal.set()
+        with self.queue.mutex:
+            self.queue.queue.clear()
+            self.queue.unfinished_tasks = 0
+            self.queue.not_full.notify()
+        self.executor.close()
+        self.run_thread.join(timeout)
+
+
+class GeneratorEnqueuer(DatasetEnqueuer):
+    """Builds a queue out of a data generator.
+
+    Used in `fit_generator`, `evaluate_generator`, `predict_generator`.
+
+    # Arguments
+        generator: a generator function which endlessly yields data
+        pickle_safe: use multiprocessing if True, otherwise threading
+        wait_time: time to sleep in-between calls to put()
+    """
+
+    def __init__(self, generator, pickle_safe=False, wait_time=0.05):
+        self.wait_time = wait_time
+        self._generator = generator
+        self._pickle_safe = pickle_safe
+        self._threads = []
+        self._stop_event = None
+        self.queue = None
+        self.wait_time = 0.0
+
+    def start(self, workers=1, max_q_size=10):
+        """Kicks off threads which add data from the generator into the queue.
+
+        # Arguments
+            workers: number of worker threads
+            max_q_size: queue size (when full, threads could block on put())
+        """
+
+        def data_generator_task():
+            while not self._stop_event.is_set():
+                try:
+                    if self._pickle_safe or self.queue.qsize() < max_q_size:
+                        generator_output = next(self._generator)
+                        self.queue.put(generator_output)
+                    else:
+                        time.sleep(self.wait_time)
+                except Exception:
+                    self._stop_event.set()
+                    raise
+
+        try:
+            if self._pickle_safe:
+                self.queue = multiprocessing.Queue(maxsize=max_q_size)
+                self._stop_event = multiprocessing.Event()
+            else:
+                self.queue = queue.Queue()
+                self._stop_event = threading.Event()
+
+            for _ in range(workers):
+                if self._pickle_safe:
+                    # Reset random seed else all children processes
+                    # share the same seed
+                    np.random.seed()
```

Comment:
This could be problematic. If you set the seed to a constant and use multi-processing then all children processes would share the same seed and you'll get batches with repeated samples.